### PR TITLE
Parse NGINX config and update NGINX instances with config context

### PR DESCRIPTION
**Pull request by dhurley**
### Proposed changes

- Parse NGINX config and update NGINX instances with config context
- Add a NginxConfigContext topic for sending NGINX config updates to other plugins

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [ ] I have read the [`CONTRIBUTING`](https://github.com/nginx/agent/blob/main/docs/CONTRIBUTING.md) document
- [ ] I have run ```make install-tools``` and have attached any dependency changes to this pull request
- [ ] If applicable, I have added tests that prove my fix is effective or that my feature works
- [ ] If applicable, I have checked that any relevant tests pass after adding my changes
- [ ] If applicable, I have updated any relevant documentation ([`README.md`](https://github.com/nginx/agent/blob/main/README.md))
- [ ] If applicable, I have tested my cross-platform changes on Ubuntu 22, Redhat 8, SUSE 15 and FreeBSD 13

### CODE REVIEW

File: internal/watcher/instance_watcher_service.go [hunk 1]
```diff
@@ -187,3 +231,61 @@ func compareInstances(oldInstances, instances []*v1.Instance) (newInstances, del
 
 	return newInstances, deletedInstances
 }
+
+func (iw *InstanceWatcherService) updateNginxInstanceRuntime(
```
Comment (aphralG) [inline internal/watcher/instance_watcher_service.go:235]: Should unit tests be added for these ? I think there was one in the resource plugin

File: internal/config/types.go [hunk 2]
```diff
@@ -125,3 +128,16 @@ type InstanceWatcher struct {
...
+
+	for _, allowedDirectory := range c.AllowedDirectories {
+		if strings.HasPrefix(directory, allowedDirectory) {
+			isDirectoryAllowed = true
+			break
+		}
+	}
+
+	return isDirectoryAllowed
+}
```
Comment (olli-holmala) [inline internal/config/types.go:143]: style: the variable and `break` aren't necessary imo.
```suggestion
func (c *Config) IsDirectoryAllowed(directory string) bool {
	for _, allowedDirectory := range c.AllowedDirectories {
		if strings.HasPrefix(directory, allowedDirectory) {
			return true
		}
	}

	return false
}
```

Comment (olli-holmala) [reply internal/config/types.go:143]: This implementation assumes that `directory` and `allowedDirectory` are both absolute paths. Will that always be the case?

File: internal/watcher/instance_watcher_service.go [hunk 3]
```diff
@@ -35,12 +38,18 @@ type (
...
-		processOperator processOperator
-		processParsers  []processParser
-		cache           []*v1.Instance
-		executer        exec.ExecInterface
+		agentConfig       *config.Config
+		processOperator   processOperator
+		processParsers    []processParser
+		nginxConfigParser nginxConfigParser
+		instanceCache     []*v1.Instance
+		nginxConfigCache  map[string]*model.NginxConfigContext
```
Comment (olli-holmala) [inline internal/watcher/instance_watcher_service.go:51]: nit:
```suggestion
		nginxConfigCache  map[string]*model.NginxConfigContext // key is instance ID
```

**Review by olli-holmala — APPROVED** [review]
Only had minor comments, looks good 👍

File: internal/config/types.go (hunk 2 above)
Comment (dhurley) [reply internal/config/types.go:143]: yes both have to be absolute paths. If they were relative paths I dont think we would have a way of finding out what directory they are in.

**Pull request merged**